modules/sdk/src/main/scala/org/tessellation/sdk/domain/snapshot/services/GlobalL0Service.scala

```diff
@@ -74,48 +168,111 @@ object GlobalL0Service {
+
+        (ordinals, List.empty[Hashed[GlobalIncrementalSnapshot]]).tailRecM {
+          case (ordinal :: nextOrdinals, snapshots) =>
+            l0GlobalSnapshotClient
+              .get(ordinal)(l0Peer)
+              .flatMap(_.toHashedWithSignatureCheck.flatMap(_.liftTo[F]))
+              .map(s => (nextOrdinals, snapshots :+ s).asLeft[List[Hashed[GlobalIncrementalSnapshot]]])
+              .handleErrorWith { e =>
+                logger
+                  .warn(e)(s"Failure pulling snapshot with ordinal=$ordinal")
+                  .map(_ => snapshots.asRight[(List[SnapshotOrdinal], List[Hashed[GlobalIncrementalSnapshot]])])
+              }
+
+          case (Nil, snapshots) =>
+            Applicative[F].pure(snapshots.asRight[(List[SnapshotOrdinal], List[Hashed[GlobalIncrementalSnapshot]])])
+        }
+      }
+
+      private case class MajoritySnapshotData(ordinal: SnapshotOrdinal, hash: Hash)
+
+      private def getLatestMajoritySnapshotData(peerIds: NonEmptyList[PeerId]): F[MajoritySnapshotData] = {
+        val maybeData =
+          for {
+            peers <- OptionT(getL0Peers(peerIds))
+            majorityOrdinal <- OptionT(getMajorityOrdinal(peers))
+            majorityHash <- OptionT(getMajorityHash(peers, majorityOrdinal))
+          } yield MajoritySnapshotData(majorityOrdinal, majorityHash)
+
+        maybeData.getOrElseF(NoMajoritySnapshotData.raiseError[F, MajoritySnapshotData])
+      }
+
+      private def getMajorityHash(peers: NonEmptyList[L0Peer], ordinal: SnapshotOrdinal): F[Option[Hash]] =
+        peers.toList
+          .parTraverseN(10)(p =>
+            attemptOption(l0GlobalSnapshotClient.getHash(ordinal).run(p), s"Unable to obtain hash for majority peer $p")
+          )
+          .map(_.flatMap(_.flatten))
+          .flatTap(hashes => logger.debug(s"** Majority Hashes $hashes"))
+          .map(selectMajorityHash(_, peers.size))
+
+      private def getPeerInTheMajority(ordinal: SnapshotOrdinal, hash: Hash, peerSet: NonEmptySet[L0Peer]): F[L0Peer] =
+        Random
+          .shuffle(List.from(peerSet.toSortedSet))
+          .tailRecM[F, L0Peer] {
+            case Nil => NoPeerAlignedWithMajority.raiseError[F, Either[List[L0Peer], L0Peer]]
+            case head :: tail =>
+              l0GlobalSnapshotClient
+                .getHash(ordinal)
+                .run(head)
+                .map {
+                  case Some(peerHash) if peerHash === hash => head.asRight[List[L0Peer]]
+                  case _                                   => tail.asLeft[L0Peer]
+                }
+          }
+
+      private def getL0Peers(peerIds: NonEmptyList[PeerId]): F[Option[NonEmptyList[L0Peer]]] =
+        peerIds
+          .traverse(globalL0ClusterStorage.getPeer)
+          .map(_.toList.flatten.toNel)
+
+      private def getMajorityOrdinal(peers: NonEmptyList[L0Peer]): F[Option[SnapshotOrdinal]] =
+        peers.toList
+          .parTraverseN(10)(p =>
```

Comment:
can you create a constant for this parallelization factor to avoid magic numbers in the code?

Reply:
@marcinwadon I was thinking of using the collection size instead of a constant. WDYT?

Reply:
won't it be too much?

Reply:
You won't have control over maximum parallel processes if you use the collection size, it will always be the collection size.